### PR TITLE
Tar reader: avoid deadlocks by synching in- and out-handlers

### PR DESCRIPTION
**Pull request by ennru**
The tar reader stage has situations where it would deadlock or get stuck in repeating the same file over and over again. This became particularly apparent when trying to extract nested tar files from within tar file in a single stream.

This PR changes the implementation to synchronize the switching of the `InHandler` and `OutHandler` for most parts, still switching the `OutHandler` when there is enough information to process from upstream.

To make theses changes I have a larger externally created tar file at hand which showed multiple flaws in the earlier tar reader implementation.

### CODE REVIEW

File: file/src/main/scala/akka/stream/alpakka/file/impl/archive/TarReaderStage.scala [hunk 1]
```diff
@@ -37,70 +37,61 @@ private[file] class TarReaderStage
...
+              new OutHandler with InHandler {
+                override def onPull(): Unit = {
+                  setHandler(flowOut, IgnoreDownstreamPull)
+                  pushSource(metadata, buffer)
+                }
+
+                // fail on upstream push
+                override def onPush(): Unit = failStage(new TarReaderException("upstream pushed unexpectedly"))
+                override def onUpstreamFinish(): Unit = setKeepGoing(true)
               }
```
Comment (seglo) [inline file/src/main/scala/akka/stream/alpakka/file/impl/archive/TarReaderStage.scala:80]: I suggest giving this handler a name like we have for the others.

File: file/src/test/java/FileProcessingUtil.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
+/*
```
Comment (seglo) [inline file/src/test/java/FileProcessingUtil.java:1]: Should we transform this into a test and/or include in the docs?

File: file/src/test/java/FileProcessingUtil.java [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+  private static final int MAX_GUNZIP_CHUNK_SIZE = 64000;
+  private static final String TARGZ_EXT = "tar.gz";
+  private static final Logger log = LoggerFactory.getLogger(FileProcessingUtil.class);
+
+  public static void main(String[] args) throws Exception {
+
+    String filenameOk = "src/test/resources/container/lbktag_eakte_0000000015.tar";
+
+    // String filename=filenameOk;
+    String filename = "file/src/test/resources/nested-sample.tar";
```
Comment (seglo) [inline file/src/test/java/FileProcessingUtil.java:34]: Should we include this artifacts?

File: file/src/test/java/docs/javadsl/NestedTarReaderTest.java [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+              } else if (targetFile.getFileName().toString().endsWith(TARGZ_EXT)) {
+                Path targetSubDir =
+                    targetFile
+                        .getParent()
+                        .resolve(
+                            Paths.get(
+                                targetFile
+                                    .getFileName()
+                                    .toString()
+                                    .substring(0, TARGZ_EXT.length() - 2)));
```
Comment (seglo) [inline file/src/test/java/docs/javadsl/NestedTarReaderTest.java:106]: Seems a little hacky. Maybe we can make this path parsing easier for the user?

**Review by seglo — COMMENTED** [review]
Looking good so far.

File: file/src/main/scala/akka/stream/alpakka/file/impl/archive/TarReaderStage.scala (hunk 1 above)
Comment (ennru) [reply file/src/main/scala/akka/stream/alpakka/file/impl/archive/TarReaderStage.scala]: 👍

File: file/src/test/java/FileProcessingUtil.java (hunk 2 above)
Comment (ennru) [reply file/src/test/java/FileProcessingUtil.java]: This shouldn't have been there. It became `NestedTarReaderTest`.

File: file/src/test/java/docs/javadsl/NestedTarReaderTest.java (hunk 4 above)
Comment (ennru) [reply file/src/test/java/docs/javadsl/NestedTarReaderTest.java:102]: This is not part of our API and I'd rather not add something for it.

**Review by seglo — APPROVED** [review]
LGTM

**Comment by seglo** [comment]
We don't have a date for the next release, but I'm hoping to release `3.0.0-M1` sometime in Feb.

**Comment by Scarysize** [comment]
Hey, sorry to chime in on this closed PR. We are also pretty anxious getting access to this fix. Either with a v3 release or a patch for v2. Any chance we can help with a back port?

**Comment by seglo** [comment]
Yes, things haven't gone as quickly as I hoped, but I plan to release `3.0.0-M1` tomorrow or over the weekend.

**Pull request merged**